table.go

```diff
@@ -53,10 +54,13 @@ type RoutingTable struct {
 	// peer in the bucket to be useful to us, failing which, we will evict
 	// it to make place for a new peer if the bucket is full
 	usefulnessGracePeriod time.Duration
+
+	df *peerdiversity.Filter
 }
 
 // NewRoutingTable creates a new routing table with a given bucketsize, local ID, and latency tolerance.
-func NewRoutingTable(bucketsize int, localID ID, latency time.Duration, m peerstore.Metrics, usefulnessGracePeriod time.Duration) (*RoutingTable, error) {
+func NewRoutingTable(bucketsize int, localID ID, latency time.Duration, m peerstore.Metrics, usefulnessGracePeriod time.Duration,
+	df *peerdiversity.Filter) (*RoutingTable, error) {
```

Comment:
@aschmahmann  After the query diversity work goes in, this struct will have the common code to:

i. Generate the grouping keys
ii. Run through the rules, whitelist blacklist etc 
iii. Persist the per-peer information to be removed when a peer is removed etc.

for both RT diversity and peer diversity. I haven't made `Filter` an interface because I don't really see multiple implementations for it.

On the other hand, `PeerIPGroupFilter ` manages the rules and the current state which will be different for the RT & Query diversity implementations.